graph2tac/loader/predict_server.py

```diff
@@ -220,8 +230,21 @@ def _enter_coq_context(self, definitions, tactics):
             if self.config.progress_bar:
                 def_clusters_for_update = tqdm.tqdm(def_clusters_for_update)
 
+            defined_nodes = set()
             for cluster in def_clusters_for_update:
                 cluster_state = self.cluster_to_graph(cluster)
+                new_defined_nodes = cluster_state.graph.nodes[:cluster_state.num_definitions]
+                used_nodes = cluster_state.graph.nodes[cluster_state.num_definitions:]
+                for n in used_nodes:
+                    assert n < prev_defined_nodes or n in defined_nodes, (
+                        f"Definition clusters out of order. "
+                        f"Attempting to compute definition embedding for node labels {new_defined_nodes} "
+                        f"({cluster_state.definition_names}) without first computing "
+                        f"the definition embedding for node label {n} used in that definition."
+                    )   
+                for n in new_defined_nodes:
+                    defined_nodes.add(n)
+
```

Comment:
This is probably the most controversial thing.  I'm checking every time that the definitions are in the correct order.  My thinking is that if we ever mess this up it could be *weeks or months* before we notice.  Also, it is difficult to make a unit test for this, since it relies on a contract with coq-tactician-reinforce.  (Also, the added time to the code is negligible.)

Reply:
The more checks the better! (Unless it causes unacceptable slowdown.)

Reply:
You could also add a check that the new defined nodes are above `prev_defined_nodes`, and were not defined before.

Reply:
I could be mistaken, but I think if we only check the new nodes, then that doesn't cover a case where a used node is somehow a bad index which is too high.  One way this could come up is if we end up repeating a node label in the same graph which defines it.  Or if somehow we end up with two definitions which mutually depend on each other.  (which shouldn't happen with our clustering).  Or if we don't get the boundary of previous nodes correct and end up skipping a bunch of new nodes leaving them undefined.  So I like this a bit better.  (But yes, to just check the definitions are in order, your approach would be sufficient.)

Reply:
Sorry, you are suggesting additional checks.  Yes, that makes sense.  I'll add them.

Reply:
Added.